File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day23OpeningTheTuringLock.scala
package advent.twenty_fifteen

import advent.common.DailyProblem
import advent.twenty_fifteen.support.{HalfRegisterOperation, IncrementRegisterOperation, JumpIfEvenOperation, JumpIfOneOperation, JumpOperation, TripleRegisterOperation, TuringOperation}
import advent.utilities.FileUtilities

import scala.annotation.tailrec

class Day23OpeningTheTuringLock(filename: String) extends DailyProblem[Map[String, Int], Map[String, Int]] {
  private val hlfPattern = "hlf ([a-z0-9]+)".r
  private val tplPattern = "tpl ([a-z0-9]+)".r
  private val incPattern = "inc ([a-z0-9]+)".r
  private val jmpPattern = "jmp ([+-][0-9]+)".r
  private val jiePattern = "jie ([a-z]+), ([+-][0-9]+)".r
  private val jioPattern = "jio ([a-z]+), ([+-][0-9]+)".r

  private def parser(line: String): TuringOperation =
    line match {
      case hlfPattern(reg) => HalfRegisterOperation(reg)
      case tplPattern(reg) => TripleRegisterOperation(reg)
      case incPattern(reg) => IncrementRegisterOperation(reg)
      case jmpPattern(off) => JumpOperation(off.toInt)
      case jiePattern(reg, off) => JumpIfEvenOperation(reg, off.toInt)
      case jioPattern(reg, off) => JumpIfOneOperation(reg, off.toInt)
      case _ => throw new IllegalStateException(s"Cannot match '$line'")
    }

  private val operations = FileUtilities.readFile(filename, parser)

  @tailrec
  private def execute(registers : Map[String, Int]) : Map[String, Int] = {
      val pc = registers("pc")
      if (pc >= operations.length)
        return registers
      val operation = operations(pc)
      execute(operation.execute(registers))
  }

  override val part1Answer: Map[String, Int] = execute(Map("a" -> 0, "b" -> 0, "pc" -> 0))
  override val part2Answer: Map[String, Int] = execute(Map("a" -> 1, "b" -> 0, "pc" -> 0))
}

File: java-advent-of-code/src/main/java/advent/twenty_fifteen/Day23OpeningTheTuringLock.java
package advent.twenty_fifteen;

import advent.common.DailyProblem;
import advent.common.Pair;
import advent.twenty_fifteen.support.TuringInstruction;
import advent.utilities.FileUtilities;

import java.util.ArrayList;
import java.util.List;

public class Day23OpeningTheTuringLock implements DailyProblem<Pair<Integer, Integer>, Pair<Integer, Integer>> {
    private final Pair<Integer, Integer> part1Answer;
    private final Pair<Integer, Integer> part2Answer;

    public Day23OpeningTheTuringLock(String filename) {
        List<TuringInstruction> program = new ArrayList<>(FileUtilities.readLines(filename, TuringInstruction::PARSE));

        int[] registers1 = executeProgram(program, 0, 0);
        this.part1Answer = new Pair<>(registers1[0], registers1[1]);

        int[] registers2 = executeProgram(program, 1, 0);
        this.part2Answer = new Pair<>(registers2[0], registers2[1]);
    }

    private int[] executeProgram(List<TuringInstruction> program, int a, int b) {
        int[] registers = new int[]{a, b};
        int ip = 0;

        while (ip < program.size()) {
            TuringInstruction instruction = program.get(ip);
            ip = instruction.execute(ip, registers);
        }
        return registers;
    }

    @Override
    public Pair<Integer, Integer> getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Pair<Integer, Integer> getPart2Answer() {
        return part2Answer;
    }
}
